src/respend/mempoolremover.cpp

```diff
@@ -3,33 +3,71 @@
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
 #include "respend/mempoolremover.h"
+#include "policy/policy.h"
+
+#include <algorithm>
 
 namespace respend {
 
-MempoolRemover::MempoolRemover(CTxMemPool& pool, std::list<CTransaction>& removed) :
-        pool(pool), removed(removed), valid(false)
+MempoolRemover::MempoolRemover(RemoveWhat toRemove,
+                               std::list<CTransaction>* removed) :
+        toRemove(toRemove), removed(removed),
+        canReplace(false), pool(nullptr), conflictingEntries(nullptr),
+        valid(false)
 {
+    if (this->removed == nullptr)
+        this->removed = &dummy;
 }
 
-bool MempoolRemover::AddOutpointConflict(
-        const COutPoint&, const CTxMemPool::txiter mempoolEntry,
-        const CTransaction& respendTx,
-        bool seenBefore, bool isEquivalent)
+bool MempoolRemover::AddOutpointConflict(const COutPoint&,
+        const CTxMemPool::txiter, const CTransaction& respendTx,
+        bool seenBefore, bool isEquivalent, bool isSICandidate)
 {
-    tx1s.insert(mempoolEntry);
+    switch (toRemove) {
+        case REMOVE_ALL:
+            return true;
+
+        case REMOVE_NON_SI:
+            if (isEquivalent)
+                return true;
+
+            canReplace = isSICandidate && respendTx.vin.size() <= 1 &&
+                         respendTx.IsImmediateRelay() &&
+                         ::GetSerializeSize(respendTx, SER_NETWORK, PROTOCOL_VERSION)
+                         <= MAX_SUPERSTANDARD_TX_SIZE;
+    }
+    return false;
+}
 
-    // Keep gathering conflicting transactions
-    return true;
+void MempoolRemover::SetupTrigger(CTxMemPool* _pool,
+        CTxMemPool::setEntries* _conflictingEntries) {
```

Comment:
How about instead of introducing a new function to feed a state, to do this instead as part of `SetValid` (SetValid then passes pool and conflicting entries). Looks to me like you only need this for the `SetValid` call. If so, I'd suggest

* Renaming `SetValid` to `OnValidTrigger` 
* Renaming `Trigger` to `OnFinishedTrigger`

Reply:
Done, this is much cleaner.